### ISSUE TITLE
`nnx.vmap` doesnt combine well with `nnx.BatchNorm`?

### ISSUE THREAD

**Issue by juliusbierk** [google/flax#4760]
This might very well be me being confused on how to use `nnx.vmap`, but I was hoping it kind of "just worked" even when using batch normalization.

However, the following example fails with an `UnexpectedTracerError`:

```python
import jax
from flax import nnx
import jax.numpy as jnp

class Model(nnx.Module):
    def __init__(self, rngs: nnx.Rngs):
        self.conv = nnx.Conv(1, 5, kernel_size=3, rngs=rngs)
        self.bn = nnx.BatchNorm(5, rngs=rngs)

    def __call__(self, x):
        x = self.bn(self.conv(x))
        return x

model = Model(rngs=nnx.Rngs(0))

# model.eval() # < if I do this the problem disappears

data = jax.random.normal(jax.random.key(0), (16, 8, 1))
print(model(data).shape)  # works
print(model(data).shape)  # works

data = jax.random.normal(jax.random.key(0), (16, 16, 8, 1))
vmapped_model = nnx.vmap(model, in_axes=1)
print(vmapped_model(data).shape)  # works
print(vmapped_model(data).shape)  # fails
```

I have also tried vmaping a function instead:

```python
data = jax.random.normal(jax.random.key(0), (16, 16, 8, 1))

def call_model(model, x):
    return model(x)

vmap_call = nnx.vmap(call_model, in_axes=(None, 1))
print(vmap_call(model, data).shape)
print(vmap_call(model, data).shape)
```

but this likewise fails.

I have found this related issue https://github.com/google/flax/issues/4061, but could not make the suggestions there work either.

Could you help me understand how `nnx.vmap` is supposed to work when using batchnorm?

**Comment by cgarciae** [comment]
Hey @juliusbierk, if you're broadcasting `BatchNorm` into `vmap` you need to reduce the BatchNorm stats in the vmap axes, to do so you need to pass `axis_name` to both BatchNorm and vmap. Here's working code:

```python
class Model(nnx.Module):
    def __init__(self, rngs: nnx.Rngs):
        self.conv = nnx.Conv(1, 5, kernel_size=3, rngs=rngs)
        self.bn = nnx.BatchNorm(5, rngs=rngs, axis_name='batch')

    def __call__(self, x):
        x = self.bn(self.conv(x))
        return x

model = Model(rngs=nnx.Rngs(0))
data = jax.random.normal(jax.random.key(0), (16, 16, 8, 1))

def call_model(model, x):
    return model(x)

vmap_call = nnx.vmap(call_model, in_axes=(None, 1), axis_name='batch')
print(vmap_call(model, data).shape)
print(vmap_call(model, data).shape)
```
Also, yeah not transform instance methods, Modules should always be explicit inputs to transformed functions.

**Comment by juliusbierk** [comment]
@cgarciae Many thanks for the fast and clear answer. Really appreciate it!
